src/components/scroll-to-anchor.tsx

```diff
@@ -0,0 +1,45 @@
+import React, { useEffect } from "react"
+
+/* This component scrolls to the element that has an ID matching
+ * the value of the ?anchor query parameter. It is designed to
+ * allow linking to a specific part of a page. This is important
+ * because web servers like IIS and Apache do not include URL
+ * fragments (i.e. the part after '#') when redirecting.
+ */
+export default function ScrollToAnchor() {
+  useEffect(() => {
+    const onPageLoad = () => {
+      // wait for browser to finish laying things out
+      setTimeout(() => {
+        const url = new URL(window.location.href)
+
+        // only do our thing if the URL doesn't have a hash already
+        if (url.hash.length === 0) {
+          const anchor = url.searchParams.get('anchor')
+
+          if (anchor) {
+            // alter the URL by converting the anchor query param into
+            // a hash
+            url.searchParams.delete('anchor')
+            url.hash = `#${anchor}`
+            window.history.pushState(null, "", url)
+            document.querySelector(`#${anchor}`)?.scrollIntoView()
+          }
+        }
+      }, 500)
+    }
+
+    // check if the page has already loaded
+    if (document.readyState === 'complete') {
+      onPageLoad()
+    } else {
+      window.addEventListener('load', onPageLoad)
+      // remove the event listener when component unmounts
+      return () => window.removeEventListener('load', onPageLoad)
+    }
+  }, [])
+
+  return (
+    <></>
+  )
+}
```

Comment:
Unless I'm missing something, I would assume that this effect would only run after the page has rendered, javascript has arrived and processed. so I don't think we need to add a listener for page load. 

Tested this out and codespaces and it seems to work well:

```suggestion
export default function ScrollToAnchor() {
  useEffect(() => {

    const url = new URL(window.location.href)

    // only do our thing if the URL doesn't have a hash already
    if (url.hash.length === 0) {
      const anchor = url.searchParams.get('anchor')

      if (anchor) {
        // alter the URL by converting the anchor query param into
        // a hash
        url.searchParams.delete('anchor')
        url.hash = `#${anchor}`
        window.history.pushState(null, "", url)
        document.querySelector(`#${anchor}`)?.scrollIntoView()
      }
    }
  }, [])

  return null
}
```

Reply:
Cool thanks for this! I tried hard to remove that 500ms delay, but the problem is the `scrollIntoView()` call happens before the browser has laid out the page, i.e. before images have loaded. When they do eventually load, the scroll position changes and the anchor is sometimes pushed below the fold.

I tried your suggestion and it seems to work about as well as any non-delay technique I was able to come up with, so let's go with it.